src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionHelper.cs

```diff
@@ -37,86 +38,158 @@ public static string GetExpressionText(LambdaExpression expression, ExpressionTe
                 return expressionText;
             }
 
-            var containsIndexers = false;
+            // Determine size of string needed and number of segments.
+            var lastIsModel = false;
+            var length = 0;
             var part = expression.Body;
-
-            // Builder to concatenate the names for property/field accessors within an expression to create a string.
-            var builder = new StringBuilder();
-
+            var segmentCount = 0;
+            var trailingMemberExpressions = 0;
             while (part != null)
             {
-                if (part.NodeType == ExpressionType.Call)
+                switch (part.NodeType)
                 {
-                    containsIndexers = true;
-                    var methodExpression = (MethodCallExpression)part;
-                    if (!IsSingleArgumentIndexer(methodExpression))
-                    {
-                        // Unsupported.
+                    case ExpressionType.Call:
+                        var methodExpression = (MethodCallExpression)part;
+                        if (IsSingleArgumentIndexer(methodExpression))
+                        {
+                            lastIsModel = false;
+                            length += 4;    // allow room for [99]
```

Comment:
Added more comments.

I am on the fence about how many characters to expect per indexer. The current approach optimizes for collections with less than 100 elements. But `containsIndexers` is the non-cached case and this penalizes dictionaries with (say) 20-`char` keys; the `StringBuilder` will be enlarged (doubled in size) at least once for each such indexer evaluation. Could detect dictionary arguments to the indexer but that seems like a penalty every indexer would pay for the 20% (or less) case.